libraries/AP_Devo_Telem/AP_Devo_Telem.cpp

```diff
@@ -0,0 +1,133 @@
+// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+
+   Inspired by work done here https://github.com/PX4/Firmware/tree/master/src/drivers/DEVO_telemetry from Stefan Rado <px4@sradonia.net>
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/* 
+   DEVO Telemetry library
+*/
+#include "AP_Devo_Telem.h"
+extern const AP_HAL::HAL& hal;
+
+DevoMPacket AP_DEVO_Telem::devoPacket = {DEVOM_SYNC_BYTE, 0};
+
+//constructor
+AP_DEVO_Telem::AP_DEVO_Telem(AP_AHRS &ahrs, const AP_BattMonitor &battery) :
+    _ahrs(ahrs),
+    _battery(battery),
+    _port(NULL),
+    _initialised_uart(false),
+    _last_frame_ms(0)
+    {}
+
+// init - perform require initialisation including detecting which protocol to use
+void AP_DEVO_Telem::init(const AP_SerialManager& serial_manager)
+{
+    
+
+    // check for DEVO_DPort
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Devo_Telem, 0))) {
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        // initialise uart
+        _port->begin(AP_SERIALMANAGER_DEVO_TELEM_BAUD, AP_SERIALMANAGER_DEVO_BUFSIZE_RX, AP_SERIALMANAGER_DEVO_BUFSIZE_TX);
+        _initialised_uart = true;   // SerialManager initialises uart for us
+
+
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_DEVO_Telem::tick, void));
+    }
+}
+
+
+uint32_t AP_DEVO_Telem::gpsDdToDmsFormat(float ddm){
+    int32_t deg = (int32_t)ddm;
+    float mm = (ddm - deg) * 60.0f;
+
+    mm = ((float)deg * 100.0f + mm) /100.0;
+
+    if ((mm < -180.0f) || (mm > 180.0f)){
+        mm = 0.0f;
+    }
+
+    return mm * 1.0e7;
+}
+
+
+/*
+  send_frames - sends updates down telemetry link 
+  should be called by main program at 1hz
+*/
+
+#define DEVO_SPEED_FACTOR 0.0194384f
+
+void AP_DEVO_Telem::send_frames(uint8_t control_mode)
+{
+    // return immediately if not initialised
+    if (!_initialised_uart) {
+        return;
+    }
+
+    const AP_GPS &gps = AP::gps();
+
+    if (gps.status() >= 3) {
+        struct Location loc = gps.location();//get gps instance 0
```

Comment:
Comment is wrong, this gets the location from primary GPS, which might even be a blend of both GPS. So either correct the comment or just remove it.

Also, why are you taking the GPS location to send instead of the one from AHRS?

Reply:
just copied from FrSky

Reply:
Well, FrSky protocol specifically defines it as GPS. I would still have used the AHRS location information, but I understand why they used GPS directly instead. That's what I was asking here, does DEVO telemetry expect this to be GPS information or just the position of the vehicle?

Reply:
These coordinates needs to find lost copter, so GPS location just inserted to GoogleMap. What to do with AHRS location in meters from start?

Reply:
AHRS can give you an absolute location, not just meters from origin or home. If you lose GPS and you have optical flow for example, you'll get a good position from AHRS but not from GPS.

Reply:
cool, but code was just copied from FrSky. And I don't think that optical flow can give a good results being above fractals, eg. forest, mountains. 

Well funny, was rejected all the proposals that are intended to combat jamming/faking GPS, but in such cases it is suggested not to rely on GPS

Reply:
> cool, but code was just copied from FrSky

Already answered that above... Also explained why getting position from AHRS is better, if you want to leave it as is, fine.

> Well funny, was rejected all the proposals that are intended to combat jamming/faking GPS

I have no idea what you are talking about.

Reply:
>Already answered that above.

I don't see an example how to get an absolute position in form lat+lon from AHRS. Only meters.

>I have no idea what you are talking about.

mostly about [this](https://github.com/ArduPilot/ardupilot/issues/7617)

Reply:
> I don't see an example how to get an absolute position in form lat+lon from AHRS. Only meters.

Here is the method: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_AHRS/AP_AHRS.h#L297
One place where it is used: 
https://github.com/ArduPilot/ardupilot/blob/99ed57f886f1e58d178f5abd12e73b5fe5133e11/APMrover2/sensors.cpp#L34-L36

> mostly about this

That's about losing GPS, not necessarily jamming/faking GPS... And it's open with quite a long technical discussion on it. So what was rejected really?

Reply:
losing GPS is definitely jamming/faking GPS in Putin's russia. All things goes bad. Problem is a reaction to this offer, seems that no one sees water

Reply:
> losing GPS is definitely jamming/faking GPS in Putin's russia

If it's faked, ArduPilot won't detect it and will continue happily flying so it's not addressed by your issue. And obviously jamming isn't the only way to lose GPS signal.

> Problem is a reaction to this offer, seems that no one sees water

Or you don't see the problems with your proposed solution...

Reply:
I just offer to allow user to choose. Describing all problems in details. So if the user is going to fly over the ocean, he does not in any case need automatic landing

Reply:
I'm not going to repeat the discussion, it was discussed and you were given the opportunity to implement it and open a PR that may or may not get accepted. You chose to not do it.

Reply:
Sadly but I don't know high level Ardupilot's code so good to do needed PR in days. But I try to implement that "jump back"